src/AppInstallerCLICore/Workflows/PinFlow.cpp

```diff
@@ -57,25 +83,32 @@ namespace AppInstaller::CLI::Workflow
     {
         auto package = context.Get<Execution::Data::Package>();
         std::vector<Pinning::Pin> pins;
-        std::set<std::string> sources;
+        std::set<std::pair<std::string, std::string>> sourcesAndIds;
 
         auto pinningIndex = context.Get<Execution::Data::PinningIndex>();
 
         auto packageVersionKeys = package->GetAvailableVersionKeys();
-        for (const auto& versionKey : packageVersionKeys)
 
+        auto installedVersion = package->GetInstalledVersion();
+        auto installedProductCodes = installedVersion->GetMultiProperty(PackageVersionMultiProperty::ProductCode);
+        auto installedPackageFamilyNames = installedVersion->GetMultiProperty(PackageVersionMultiProperty::PackageFamilyName);
+
+        for (const auto& versionKey : packageVersionKeys)
         {
             auto availableVersion = package->GetAvailableVersion(versionKey);
-            Pinning::PinKey pinKey{
-            availableVersion->GetProperty(PackageVersionProperty::Id).get(),
-                availableVersion->GetProperty(PackageVersionProperty::SourceIdentifier).get() };
+            const auto availablePackageId = availableVersion->GetProperty(PackageVersionProperty::Id).get();
+            const auto availableSourceId = availableVersion->GetProperty(PackageVersionProperty::SourceIdentifier).get();
+            const auto pinKeys = Pinning::GetPinKeysForAvailablePackage(availablePackageId, availableSourceId, installedProductCodes, installedPackageFamilyNames);
 
-            if (sources.insert(pinKey.SourceId).second)
+            for (const auto& pinKey : pinKeys)
             {
-                auto pin = pinningIndex->GetPin(pinKey);
-                if (pin)
+                if (sourcesAndIds.emplace(pinKey.SourceId, pinKey.ExtraId).second)

```

Comment:
If the purpose is to avoid duplicate pins, should we just make pins an std::set and not worry about 2 specific fields in pinkey?